sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -17,109 +22,279 @@
 )
 def test_default_request(metaestimator):
     # Check that by default request is empty and the right type
-    assert_request_is_empty(metaestimator.get_metadata_routing())
-    assert isinstance(metaestimator.get_metadata_routing(), MetadataRouter)
+    cls = metaestimator["metaestimator"]
+    estimator = metaestimator["estimator"]()
+    estimator_name = metaestimator["estimator_name"]
+    instance = cls(**{estimator_name: estimator})
+    assert_request_is_empty(instance.get_metadata_routing())
+    assert isinstance(instance.get_metadata_routing(), MetadataRouter)
 
 
 @pytest.mark.parametrize(
-    "MultiOutput, Estimator",
-    [
-        (MultiOutputClassifier, ConsumingClassifier),
-        (MultiOutputRegressor, ConsumingRegressor),
-    ],
-    ids=["Classifier", "Regressor"],
+    "metaestimator",
+    METAESTIMATORS,
+    ids=[str(row["metaestimator"].__class__.__name__) for row in METAESTIMATORS],
 )
-def test_multioutput_metadata_routing(MultiOutput, Estimator):
-    # Check routing of metadata
-    metaest = MultiOutput(Estimator())
-    warn_msg = (
-        "You are passing metadata for which the request values are not explicitly"
-        " set: sample_weight, metadata."
-    )
-    with pytest.warns(FutureWarning, match=(warn_msg)):
-        metaest.fit(X, y_multi, sample_weight=sample_weight, metadata=metadata)
-        check_recorded_metadata(
-            metaest.estimators_[0],
-            "fit",
-            sample_weight=sample_weight,
-            metadata=metadata,
+def test_warning_for_indicated_methods(metaestimator):
+    # Check that the indicated methods give a warning
+    # TODO: After deprecation period, always error
+    cls = metaestimator["metaestimator"]
+    registry = _Registry()
+    estimator = metaestimator["estimator"](registry=registry)
+    estimator_name = metaestimator["estimator_name"]
+    instance = cls(**{estimator_name: estimator})
+    X = metaestimator["X"]
+    y = metaestimator["y"]
+    routing_methods = metaestimator["routing_methods"]
+
+    for method_name in routing_methods:
+        if not metaestimator.get(method_name, {}).get("warns_on"):
+            # this method is not expected to warn
+            continue
+
+        warn_msg = (
+            "You are passing metadata for which the request values are not explicitly "
+            "set: sample_weight, metadata. From version 1.4 this results in the "
+            "following error: [sample_weight, metadata] are passed but are not "
+            "explicitly set as requested or not for "
+            f"{estimator.__class__.__name__}.{method_name}"
         )
+        with pytest.warns(FutureWarning, match=re.escape(warn_msg)):
+            method = getattr(instance, method_name)
+            method(X, y, sample_weight=sample_weight, metadata=metadata)
+
+        if metaestimator.get("preserves_metadata", True):
+            check_recorded_metadata(
+                registry[-1],
```

Comment:
shouldn't we check all members of registry, to make sure metadata is correctly passed everywhere?

Reply:
Yes, we can do that, but I'll clear the list between for each method, since there could be cases where the check differs for different methods (at least I think so).